File: Projekt PO Swiat cpp/Antylopa.cpp
#include "Antylopa.h"


Antylopa::Antylopa(const int polozenieX, const int polozenieY, const int wiek) {
	setNazwa("Antylopa");
	setSymbol('A');
	setSila(4);
	setInicjatywa(4);
	setWiek(wiek);
	setPolozenieX(polozenieX);
	setPolozenieY(polozenieY);
	setNiesmiertelnosc(false);
}


Antylopa::Antylopa(const Organizm& inny) {
	kopiujObiekt(inny);
}


Organizm* Antylopa::stworzNowySklonowanyObiekt() {
	return new Antylopa(*this);
}


void Antylopa::akcja(Swiat& swiat) {
	int wymiarMapyX = swiat.getWymiarMapyX(), wymiarMapyY = swiat.getWymiarMapyY(), polozenieX = getPolozenieX(), polozenieY = getPolozenieY();
	vector<int> mozliweMiejscaX;
	vector<int> mozliweMiejscaY;

	for (int y = polozenieY - 2; y <= polozenieY + 2; y += 2)
	{
		if (y < 0 || y >= wymiarMapyY) continue;
		for (int x = polozenieX - 2; x <= polozenieX + 2; x+=2)
		{
			if (x < 0 || x >= wymiarMapyX) continue;
			if (x == polozenieX && y == polozenieY) continue;
			mozliweMiejscaX.push_back(x);
			mozliweMiejscaY.push_back(y);
		}
	}

	if (mozliweMiejscaX.size() == 0) return;

	int ruch = rand() % mozliweMiejscaX.size();
	int nowyX = mozliweMiejscaX[ruch], nowyY = mozliweMiejscaY[ruch];

	if (swiat.getOrganizm(nowyX, nowyY) == nullptr) {
		swiat.usunOrganizm(this, polozenieX, polozenieY);
		setPolozenieX(nowyX);
		setPolozenieY(nowyY);
		swiat.dodajOrganizm(this, nowyX, nowyY);
	}

	else if (swiat.getOrganizm(nowyX, nowyY)->getSymbol() == swiat.getOrganizm(polozenieX, polozenieY)->getSymbol()) {
		Organizm* nowaAntylopa = new Antylopa(*this);
		Zwierze::kolizja(swiat, *nowaAntylopa);
	}

	else {
		swiat.getOrganizm(nowyX, nowyY)->kolizja(swiat, *this);
	}


}


void Antylopa::kolizja(Swiat& swiat, Organizm& atakujacy) {
	int polozenieX = getPolozenieX(), polozenieY = getPolozenieY(), unik = rand() % 2, polozenieXAtak = atakujacy.getPolozenieX(), polozenieYAtak = atakujacy.getPolozenieY();
	
	if (unik == 1) { 
		if (unikAntylopy(swiat, atakujacy)) return;
	}

	bool czyPrzetrwal = czyOdbilAtak(atakujacy, *this);
	if (czyPrzetrwal) {
		swiat.wypiszWiadomosc("Antylopa zabija " + atakujacy.getNazwa());
		atakujacy.setWiek(NIEZYWY_ORGANIZM);
		swiat.usunOrganizm(&atakujacy, polozenieXAtak, polozenieYAtak);
	}

	else {
		setWiek(NIEZYWY_ORGANIZM);
		swiat.usunOrganizm(this, polozenieX, polozenieY);
		atakujacy.setPolozenieX(polozenieX);
		atakujacy.setPolozenieY(polozenieY);
		swiat.dodajOrganizm(&atakujacy, polozenieX, polozenieY);
		swiat.usunOrganizm(&atakujacy, polozenieXAtak, polozenieYAtak);
		swiat.wypiszWiadomosc(atakujacy.getNazwa() + " zabija Antylopa");
	}
}


bool Antylopa::unikAntylopy(Swiat& swiat, Organizm& atakujacy) {
	int polozenieX = getPolozenieX(), polozenieY = getPolozenieY(), polozenieXAtak = atakujacy.getPolozenieX(), polozenieYAtak = atakujacy.getPolozenieY();
	for (int y = polozenieY - 1; y <= polozenieY + 1; y++)
	{
		if (y < 0 || y >= swiat.getWymiarMapyY()) continue;
		for (int x = polozenieX - 1; x <= polozenieX + 1; x++)
		{
			if (x < 0 || x >= swiat.getWymiarMapyX()) continue;
			if (swiat.getOrganizm(x, y) == nullptr) {
				swiat.usunOrganizm(this, polozenieX, polozenieY);
				setPolozenieX(x);
				setPolozenieY(y);
				atakujacy.setPolozenieX(polozenieX);
				atakujacy.setPolozenieY(polozenieY);
				swiat.dodajOrganizm(this, x, y);
				swiat.dodajOrganizm(&atakujacy, polozenieX, polozenieY);
				swiat.usunOrganizm(&atakujacy, polozenieXAtak, polozenieYAtak);
				swiat.wypiszWiadomosc("Antylopa unika ataku " + atakujacy.getNazwa());
				return true;
			}
		}
	}
	return false;
}


char Antylopa::rysowanie() {
	textcolor(BROWN);
	return getSymbol();
}


Antylopa::~Antylopa() {}

File: wirtualny_swiat_python/Antylopa.py
import random
from Zwierze import Zwierze


class Antylopa(Zwierze):
    def __init__(self, x, y, wiek):
        super().__init__(x, y, wiek)
        self.setNazwa("Antylopa")
        self.setSymbol('A')
        self.setSila(4)
        self.setInicjatywa(4)
        self.setNiesmiertelnosc(False)

    def stworzNowySklonowanyObiekt(self):
        return Antylopa(0, 0, 0)

    def akcja(self, swiat):
        wymiarMapyX = swiat.getWymiarMapyX()
        wymiarMapyY = swiat.getWymiarMapyY()
        mozliweMiejscaX = []
        mozliweMiejscaY = []

        for y in range(self.polozenieY - 2, self.polozenieY + 3):
            if y < 0 or y >= wymiarMapyY:
                continue
            for x in range(self.polozenieX - 2, self.polozenieX + 3):
                if x < 0 or x >= wymiarMapyX:
                    continue
                if x == self.polozenieX and y == self.polozenieY:
                    continue
                mozliweMiejscaX.append(x)
                mozliweMiejscaY.append(y)

        if len(mozliweMiejscaX) == 0:
            return

        ruch = random.randint(0, len(mozliweMiejscaX) - 1)
        nowyX = mozliweMiejscaX[ruch]
        nowyY = mozliweMiejscaY[ruch]

        if swiat.getOrganizm(nowyX, nowyY) is None:
            swiat.usunOrganizm(self, self.polozenieX, self.polozenieY)
            self.setPolozenieX(nowyX)
            self.setPolozenieY(nowyY)
            swiat.dodajOrganizm(self, nowyX, nowyY)

        elif swiat.getOrganizm(nowyX, nowyY).getSymbol() == self.symbol:
            noweZwierze = self.stworzNowySklonowanyObiekt()
            super().kolizja(swiat, noweZwierze)

        else:
            swiat.getOrganizm(nowyX, nowyY).kolizja(swiat, self)

    def kolizja(self, swiat, atakujacy):
        unik = random.randint(0, 1)
        if unik == 1:
            if self.unikAntylopy(swiat, atakujacy):
                return
        super().kolizja(swiat, atakujacy)

    def unikAntylopy(self, swiat, atakujacy):
        polozenieXAtak = atakujacy.getPolozenieX()
        polozenieYAtak = atakujacy.getPolozenieY()
        for y in range(self.polozenieY - 1, self.polozenieY + 2):
            if y < 0 or y >= swiat.getWymiarMapyY():
                continue
            for x in range(self.polozenieX - 1, self.polozenieX + 2):
                if x < 0 or x >= swiat.getWymiarMapyX():
                    continue
                if swiat.getOrganizm(x, y) is None:
                    swiat.usunOrganizm(self, self.polozenieX, self.polozenieY)
                    swiat.usunOrganizm(atakujacy, polozenieXAtak, polozenieYAtak)
                    atakujacy.setPolozenieX(self.polozenieX)
                    atakujacy.setPolozenieY(self.polozenieY)
                    swiat.dodajOrganizm(atakujacy, self.polozenieX, self.polozenieY)
                    self.setPolozenieX(x)
                    self.setPolozenieY(y)
                    swiat.dodajOrganizm(self, x, y)
                    swiat.wypiszWiadomosc("Antylopa unika ataku " + atakujacy.getNazwa()
                                          + self.wypiszPolozenie(self.polozenieX, self.polozenieY))
                    return True
        return False

    def rysowanie(self):
        return "sienna"
